### PR TITLE
WebMethod improvements

### PR DESCRIPTION
**Pull request by kohsuke**
To help create better REST APIs with POJOs,

* Define a series of HTTP method annotations like `DELETE` and `GET` that lets you write different methods to handle different verbs for the same URL.
* Allow `@WebMethod` on any method, not just those starting with the `do` prefix

### CODE REVIEW

**Comment by kohsuke** [comment]
@reviewbybees though I'd like to hear from everyone, not just fellow bees.

File: core/src/main/java/org/kohsuke/stapler/Function.java [hunk 1]
```diff
@@ -94,9 +94,11 @@
     boolean bindAndInvokeAndServeResponse(Object node, RequestImpl req, ResponseImpl rsp, Object... headArgs) throws IllegalAccessException, InvocationTargetException, ServletException, IOException {
         try {
             Object r = bindAndInvoke(node, req, rsp, headArgs);
-            if (getReturnType()!=void.class)
-                renderResponse(req,rsp,node, r);
+            if (getReturnType() != void.class)
+                renderResponse(req, rsp, node, r);
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: Thanks for these changes :+1:

File: core/src/main/java/org/kohsuke/stapler/FunctionList.java [hunk 2]
```diff
@@ -110,7 +120,19 @@ public boolean keep(Function m) {
...
+    }
+
+    /**
+     * Returns {@link Function}s that are either explicitly {@link WebMethod} or
+     * implicitly so (by having its name start with 'do')
+     */
+    public FunctionList webMethods() {
+        return filter(new Filter() {
+            public boolean keep(Function m) {
+                return m.getName().startsWith("do") || m.getAnnotation(WebMethod.class)!=null;
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/FunctionList.java]: `m.getAnnotation(WebMethod.class) != null`

File: core/src/main/java/org/kohsuke/stapler/MetaClass.java [hunk 3]
```diff
@@ -105,8 +105,8 @@
         this.dispatchers.clear();
         ClassDescriptor node = new ClassDescriptor(clazz,null/*TODO:support wrappers*/);
 
-        // check action <obj>.do<token>(...)
-        for( final Function f : node.methods.prefix("do") ) {
+        // check action <obj>.do<token>(...) and other WebMethods
+        for( final Function f : node.methods.webMethods() ) {
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/MetaClass.java]: `for (final ...`

File: core/src/main/java/org/kohsuke/stapler/MetaClass.java [hunk 4]
```diff
@@ -187,26 +187,19 @@ public String toString() {
...
-                    }
-                    public String toString() {
-                        return String.format("%1$s() for url=/%2$s/...",f.getQualifiedName(),name);
-                    }
-                });
-            }
+            dispatchers.add(new NameBasedDispatcher(name) {
+                public boolean doDispatch(RequestImpl req, ResponseImpl rsp, Object node) throws IOException, ServletException, IllegalAccessException, InvocationTargetException {
+                    if(traceable())
+                        traceEval(req,rsp,node,f.getName()+"()");
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/MetaClass.java]:
```
if (traceable()) {
    traceEval(req, rsp, node, f.getName() + "()");
}
```

File: core/src/main/java/org/kohsuke/stapler/interceptor/Interceptor.java [hunk 5]
```diff
@@ -45,6 +46,8 @@ public void setTarget(Function target) {
      * @return
      *      Return value from the method.
      * @throws InvocationTargetException if you want to send e.g. something from {@link HttpResponses}
+     * @throws CancelRequestHandlingException
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/interceptor/Interceptor.java]: `CancelRequestHandlingException` or `IllegalAccessException`?

File: core/src/main/java/org/kohsuke/stapler/verb/HttpVerbInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            throw new CancelRequestHandlingException();
+    }
+
+    private boolean matches(StaplerRequest request) {
+        String method = request.getMethod();
+
+        for (Annotation a : target.getAnnotations()) {
+            Class<? extends Annotation> t = a.annotationType();
+            InterceptorAnnotation ia = t.getAnnotation(InterceptorAnnotation.class);
+            if (ia !=null && ia.value()==HttpVerbInterceptor.class) {
```
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/verb/HttpVerbInterceptor.java]: `ia.value() == HttpVerbInterceptor.class`

Really I'm a human, not a coding style bot :panda_face:

**Comment by recena** [comment]
@kohsuke I tried to verify this PR in local, but there are conflicts. Do you plan to solve this?

File: core/src/main/java/org/kohsuke/stapler/Function.java (hunk 1 above)
Comment (kohsuke) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: Actually unintentional, I got some more IDE settings to tweak, it looks like

File: core/src/main/java/org/kohsuke/stapler/interceptor/Interceptor.java (hunk 5 above)
Comment (kohsuke) [inline core/src/main/java/org/kohsuke/stapler/interceptor/Interceptor.java]: `CancelRequestHandlingException` as stated in javadoc. Or maybe I'm missing your point.

File: core/src/main/java/org/kohsuke/stapler/Function.java (hunk 1 above)
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: intentional or not, :+1: for your new settings.

**Comment by jglick** [comment]
Somewhere—but how do I find it?—I wrote up an argument that `@WebMethod` should become mandatory, so that we would drop the magical handling of `do*` methods. I had a proposal for how to make this binary compatible, I think using an annotation processor. Also that `@RequirePOST` would become an attribute of `@WebMethod`, or simply the default.

**Comment by jglick** [comment]
Ah! In #32.

File: core/src/main/java/org/kohsuke/stapler/Function.java (hunk 1 above)
Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: :-1: on gratuitous reformatting, which makes the diff harder to review.

**Comment by jglick** [comment]
Specifically I advocate that

* you _must_ place `@WebMethod` on any method you expect to be treated in this way—no more surprise exposure of security-sensitive code to the world
  * `name` should be mandatory: a simple Java rename refactoring should not affect behavior
* you _must_ select one or more HTTP verbs—no more forgetting to disallow `GET` on mutators

I think these changes would make for more explicit and safer code. They are not source-compatible but could be made binary-compatible with a little more effort.

File: core/src/main/java/org/kohsuke/stapler/Function.java (hunk 1 above)
Comment (recena) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: come on @jglick !!! you know perfectly that these little changes improve the readability of the source code.

I know also, you can review this PR even if it has obfuscated :dizzy_face:

Comment (jglick) [inline core/src/main/java/org/kohsuke/stapler/Function.java]: > these little changes improve the readability of the source code

Yes, very slightly, but they significantly _harm_ the readability of the PR. If you want to do a bulk reformat in a separate PR (and introduce a “diff wall” past which merging & cherry-picking becomes impractical), that is another matter.

**Comment by kohsuke** [comment]
I think Jesse is suggesting a change that goes much beyond the scope of this PR.

What this PR does is that it lets us approach that change better than what's stated in issue #32 (namely by having explicit HTTP verb annotations, POST need not be treated any differently from other verbs.)

I'll capture that in a separate ticket for a future consideration, though I don't think I want to make configuration explicit like that.

**Comment by jglick** [comment]
> I think Jesse is suggesting a change that goes much beyond the scope of this PR.

To some extent, yes, but merging this takes the API in a direction from which it might be harder to recover than where we started.
